fpga_interchange/populate_chip_info.py

```diff
@@ -0,0 +1,487 @@
+            primary_site_type = device.device_resource_capnp.siteTypeList[site_type_index]
+
+            self.add_site_type(device, site_type_in_tile_type, site_in_type_index, site_type_index, site_variant)
+
+            for site_variant, (alt_site_type_index, _) in enumerate(zip(primary_site_type.altSiteTypes, site_type_in_tile_type.altPinsToPrimaryPins)):
+                self.add_site_type(device, site_type_in_tile_type, site_in_type_index, alt_site_type_index, site_variant, primary_site_type)
+
+    def add_wire(self, wire):
+        wire_index = len(self.wires)
+        self.wires.append(wire)
+
+        return wire_index
+
+    def add_pip_common(self, flat_pip):
+        pip_index = len(self.pips)
+
+        self.pips.append(flat_pip)
+
+        self.wires[flat_pip.src_index].pips_downhill.append(pip_index)
+        self.wires[flat_pip.dst_index].pips_uphill.append(pip_index)
+
+    def add_tile_pip(self, tile_pip_index, src_wire, dst_wire):
+        assert self.wires[src_wire].type == FlattenedWireType.TILE_WIRE
+        assert self.wires[dst_wire].type == FlattenedWireType.TILE_WIRE
+
+        flat_pip = FlattenedPip(
+                type=FlattenedPipType.TILE_PIP,
+                src_index=src_wire,
+                dst_index=dst_wire,
+                site_index=None,
+                pip_index=tile_pip_index)
+
+        self.add_pip_common(flat_pip)
+
+    def add_site_type(self, device, site_type_in_tile_type, site_in_type_index, site_type_index, site_variant, primary_site_type=None):
+        if site_variant == -1:
+            assert primary_site_type is None
+        else:
+            assert primary_site_type is not None
+
+        site_index = len(self.sites)
+
+        bel_to_bel_index = {}
+        bel_pin_to_site_wire_index = {}
+        bel_pin_index_to_bel_index = {}
+
+        site_type = device.device_resource_capnp.siteTypeList[site_type_index]
+
+        self.sites.append(FlattenedSite(
+            site_in_type_index=site_in_type_index,
+            site_type_index=site_type_index,
+            site_type=site_type,
+            site_variant=site_variant,
+            bel_to_bel_index=bel_to_bel_index,
+            bel_pin_to_site_wire_index=bel_pin_to_site_wire_index,
+            bel_pin_index_to_bel_index=bel_pin_index_to_bel_index
+            ))
+
+        # Add site wires
+        for idx, site_wire in enumerate(site_type.siteWires):
+            wire_name = device.strs[site_wire.name]
+            flat_wire = FlattenedWire(
+                    type=FlattenedWireType.SITE_WIRE,
+                    name=wire_name,
+                    wire_index=idx,
+                    site_index=site_index)
+
+            site_wire_index = self.add_wire(flat_wire)
+            for pin in site_wire.pins:
+                assert pin not in bel_pin_to_site_wire_index
+                bel_pin_to_site_wire_index[pin] = site_wire_index
+
+        # Add BELs
+        for bel_idx, bel in enumerate(site_type.bels):
```

Comment:
I'm not sure if there is any need to import site port bels as bels; given pips are being inserted for the tile/site boundary already and are more suited for this kind of purpose inside nextpnr. They won't do any harm, but will potentially add clutter and marginally slow things down.

It's also arguable as to whether routing bels are actually needed as bels either - again importing these as just pips might be better - I know nextpnr-xilinx historically created bels for these but I don't remember if they were ever actually used as bels. All the needed information could probably be tagged to the pips.

Reply:
> It's also arguable as to whether routing bels are actually needed as bels either - again importing these as just pips might be better - I know nextpnr-xilinx historically created bels for these but I don't remember if they were ever actually used as bels. All the needed information could probably be tagged to the pips.

I actually ended up using both the routing and site pin BELs, so for now they are getting emitted.  If it is a problem to have these BELs in the chip data, I could put the routing and site pin BELs at the end of the tile type bel_data array so they are easily skipped.

Reply:
That's fair enough - the only real problems I was thinking were related to debugging and putting them at the end would make it a bit clearer what was going on